File: Juan/P3/Ruby/Napakalaki/lib/Napakalaki.rb
#encoding: utf-8

#Versión 3.0
require 'singleton'
require 'Player.rb'
require 'CardDealer.rb'
require 'Dice.rb'
require 'singleton'


module NapakalakiGame


class Napakalaki 
   include Singleton
   
    #@@instance = nil
#    Constructor

#    def initialize(currentPlayer,players,dealer,currentMonster,unusedMonsters)
#        @currentPlayer = currentPlayer
#        @players = players
#        @dealer = dealer
#        @currentMonster = currentMonster
#        @unusedMonsters = unusedMonsters
#    end


    def initialize()
        @currentPlayer 
        @players 
        @dealer = CardDealer.instance
        @currentMonster  
    end

    attr_accessor :currentPlayer , :player, :dealer, :currentMonster
    
    ## getCurrentPlayer
    ## getCurrentMonster
    #attr_reader :currentPlayer
    #attr_reader :currentMonster
    
    #private_class_method :new
    
#    Inicializa el array de jugadores que contiene Napakalaki, creando tantos jugadores como
#    elementos haya en names, que es el array de String que contiene el nombre de losaa
#    jugadores.
   
    def initPlayers(names)
         @players = Array.new
           names.each do |n|
           @players <<  Player.new(n)
           
          end
         
    end
    
    public
    def getCurrentPlayer
        
        return @currentPlayer
    end

    private 
    def nextPlayer()
        if(@currentPlayer == nil)
            numero = rand(@players.size)
            puts @players.size
            puts @players.fetch(0)
            puts @players.fetch(1)
            @currentPlayer = @players[numero]
           
            aux=Array.new
            
            aux << @currentPlayer
            
            @players.each do |p|
                if ( p != @currentPlayer)
                    aux << p
                end
            end
            @players= aux
            
            return @currentPlayer
            
        else
            i=0
            while( i< @players.size)
                if(@currentPlayer == @player[i])
                    if(i==(@players.size -1))
                        @currentPlayer = @players[0]
                    else
                        @currentPlayer = @players[i+1]
                    end
                end
                i = i+1
            end
            
        end
        return @currentPlayer
    end
    

    public
    def getCurrentPlayer
        
        return @currentPlayer
    end
    
    def getCurrentMonster
        return @currentMonster
    end
    
    private
    def nextTurnAllowed()
        sig = false
        
        #if(@currentPlayer.validState)
        if @currentPlayer == nil
            sig=true
        end
        
        return sig
    end
    
    
    private
    def setEnemies()
        enemigo =Player.new(@players[0])
        #enemigo = @players[0]
        
        @players.each do |p|
            
            loop do 
                 numero = rand(@players.size)
                  
                 enemigo.setEnemy(@players[numero])
                 
            break if (enemigo != p)
            end
        end 
    end
    
#    private 
#    def self.getInstance()
#        @@instance = Napakalaki.instance
#    end
#   

    public
    def developCombat()
        combatResult = @currentPlayer.combat(@currentMonster)
        
        return combatResult
    end
    

    
    def discardVisibleTreasures (treasures)
        treasures.each do |t|
            @currentPlayer.discardVisibleTreasure(t)
            @dealer.giveTreasureBack(t)
        end
    end
    
    def discarHiddenTreasures(treasures)
        treasures.each do |t|
            @currentPlayer.discardHiddenTreasure(t)
            @dealer.giveTreasureBack(t)
        end
    end
    
<<<<<<< HEAD
    def makeTreasureVisible(treasures)

    
=======
   

    def makeTreasureVisible(treasures)

>>>>>>> origin/master
        treasures.each do |t|
            @currentPlayer.discardHiddenTreasure(t)
            @dealer.giveTreasureBack(t)
        end
        
    end
    
    def initGame(players)
        initPlayers(players)
        setEnemies()

        @dealer.initCards
         nextTurn()
       # @currentPlayer.initTreasures
       
        
    end
    

    def nextTurn()
        stateOK = nextTurnAllowed
        
        if(stateOK == true)
            @currentMonster = @dealer.nextMonster
            @currentPlayer = nextPlayer
            #dead = @currentPlayer.isDead()
            
            #if(dead == true)
                @currentPlayer.initTreasures
            #end
        end
        
        return stateOK
    end
    
    def endOfGame (result)
        
       final = false
       if (result.eql?([CombatResult::WINGAME]))
            final = true
            
       end
            return final
   end
   
    
    
end

end

File: Juan/P3/Java/Napakalaki/src/napakalaki/Napakalaki.java
package napakalaki;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author juane
 * @version 3.1 clase singleton
 */
public class Napakalaki {

    private static final Napakalaki instance = null;

    /* Relaciones con las clases */
    private Monster currentMonster; // Relación con los monstruos    
    private CardDealer dealer; // Relación con CardDealer
    private Player currentPlayer; // Relación con Player
    private ArrayList<Player> players = new ArrayList(); // Relación con Player

    /**
     * Constructor privado
     */
    private Napakalaki() {

    }

    /**
     * Método initPlayers(ArrayList<String> names
     *
     * @param names
     */
    private void initPlayers(ArrayList<String> names) {
        for (String n : names) {
            this.players.add(new Player(n));
        }
    }

    /**
     * Método nextPlayer Decide qué jugador es el siguiente en jugar.
     *
     * @return
     */
    private Player nextPlayer() {

        if (currentPlayer == null) {
            Random rand = new Random();
            int numero = rand.nextInt(players.size());

            currentPlayer = players.get(numero);
            ArrayList<Player> aux = new ArrayList();

            aux.add(currentPlayer);

            for (Player p : players) {
                if (p != currentPlayer) {
                    aux.add(p);
                }
            }
            players = aux;

            return currentPlayer;
            
        } else {
            for (int i = 0; i < players.size(); i++) {
                if (currentPlayer == players.get(i)) {
                    if (i == players.size() - 1) ///////si falla mirara aquí
                    {
                        currentPlayer = players.get(0);
                    } else {
                        currentPlayer = players.get(i + 1);
                    }
                }
            }

            return currentPlayer;
        }
    }

    /**
     * Método nextTurnAllowed()
     *
     * @return
     */
    private boolean nextTurnAllowed() {

        boolean next = false;

        if (this.currentPlayer.validState()) {
            next = true;
        }

        return next;
    }

    /**
     * Método setEnemies()
     */
    private void setEnemies() {

        Player enemigo = players.get(0);

        Random rand = new Random();
        for (Player p : players) {
            do {
                int numero = rand.nextInt(players.size());
                enemigo.setEnemy(players.get(numero));

            } while (enemigo.enemy == p);
        }

    }

    /**
     * Método getInstance()
     *
     * @return
     */
    public static Napakalaki getInstance() {
        return instance;
    }

    /**
     * Método developCombat()
     *
     * @return
     */
    public CombatResult developCombat() {

        CombatResult combatResult = this.currentPlayer.combat(this.currentMonster);

        return combatResult;
    }

    /**
     * Método discardVisibleTreasures(ArrayList<Treasure> treasures)
     *
     * @param treasures
     */
    public void discardVisibleTreasures(ArrayList<Treasure> treasures) {
        for (Treasure t : treasures) {
            this.currentPlayer.discardVisibleTreasure(t);

            this.dealer.giveTreasureBack(t);
        }
    }

    /**
     * Método discardHiddenTreasures(ArrayList<Treasure> treasures)
     *
     * @param treasures
     */
    public void discardHiddenTreasures(ArrayList<Treasure> treasures) {
        for (Treasure t : treasures) {
            this.currentPlayer.discardHiddenTreasure(t);

            this.dealer.giveTreasureBack(t);
        }
    }

    /**
     * Método makeTreasuresVisible(ArrayList<Treasure> treasures)
     *
     * @param treasures
     */
    public void makeTreasuresVisible(ArrayList<Treasure> treasures) {

        for (Treasure t : treasures) {
            this.currentPlayer.makeTreasureVisible(t);
        }
    }

    /**
     * Método initGame(ArrayList<String> players)
     *
     * @param players
     */
    public void initGame(ArrayList<String> players) {
        this.initPlayers(players);
        this.setEnemies();
        this.dealer.initCards();
        this.nextTurn();

    }

    /**
     * Método getCurrentPlayer()
     *
     * @return
     */
    public Player getCurrentPlayer() {

        return this.currentPlayer; //cambiar
    }

    /**
     * Método getCurrentMonster()
     *
     * @return
     */
    public Monster getCurrentMonster() {
        return this.currentMonster;
    }

    /**
     * Método nextTurn()
     *
     * @return
     */
    public boolean nextTurn() {
        //MIRAR
        boolean stateOK;
        stateOK = this.nextTurnAllowed();  
             
        
        if (stateOK) {
            boolean dead;
            this.currentMonster = this.dealer.nextMonster();
            this.currentPlayer = this.nextPlayer();
            dead = this.currentPlayer.isDead();

            if (dead) {
                this.currentPlayer.initTreasures();
            }
        }
        stateOK = this.currentPlayer.validState(); //PODRIA SER ASÍ
        return stateOK;
    }

    /**
     * Método endOfGame(CombatResult result)
     *
     * @param result
     * @return
     */
    public boolean endOfGame(CombatResult result) {

        boolean end = false;
        if (result.equals(CombatResult.WINGAME)) {
            end = true;
        }

        return end;
    }
}
